test/mpi/test_mpi_cuda.c

```diff
@@ -0,0 +1,668 @@
+
+    CUDA_CALL(cuDevicePrimaryCtxRetain(&cu_ctx, cu_dev));
+    CUDA_CALL(cuCtxPushCurrent(cu_ctx));
+}
+
+static void pop_and_release_primary_context(CUdevice cu_dev)
+{
+    CUDA_CALL(cuCtxPopCurrent(NULL));
+    CUDA_CALL(cuDevicePrimaryCtxRelease(cu_dev));
+}
+
+static alloc_mem_t alloc_mem_alloc(size_t size)
+{
+    alloc_mem_t alloc_mem = {};
+
+    CUDA_CALL(cuMemAlloc(&alloc_mem.ptr, size));
+
+    return alloc_mem;
+}
+
+static void free_mem_alloc(alloc_mem_t *alloc_mem)
+{
+    CUDA_CALL(cuMemFree(alloc_mem->ptr));
+}
+
+static alloc_mem_t alloc_mem_alloc_managed(size_t size)
+{
+    alloc_mem_t alloc_mem = {};
+
+    CUDA_CALL(cuMemAllocManaged(&alloc_mem.ptr, size, CU_MEM_ATTACH_GLOBAL));
+
+    return alloc_mem;
+}
+
+static alloc_mem_t alloc_vmm_type(size_t size, unsigned handle_type)
+{
+    CUmemAllocationProp prop    = {};
+    size_t granularity          = 0;
+    CUmemAccessDesc access_desc = {};
+    alloc_mem_t alloc_mem       = {};
+    CUdevice cu_dev;
+    CUmemGenericAllocationHandle handle;
+    CUresult result;
+    CUdeviceptr ptr;
+
+    CUDA_CALL(cuCtxGetDevice(&cu_dev));
+
+    prop.type          = CU_MEM_ALLOCATION_TYPE_PINNED;
+    prop.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+    prop.location.id   = cu_dev;
+    if (handle_type != 0) {
+        prop.requestedHandleTypes = (CUmemAllocationHandleType)handle_type;
+    }
+
+    prop.allocFlags.gpuDirectRDMACapable = 1;
+
+    CUDA_CALL(cuMemGetAllocationGranularity(&granularity, &prop,
+                                            CU_MEM_ALLOC_GRANULARITY_MINIMUM));
+    if ((size % granularity) != 0) {
+        size = ((size / granularity) + 1) * granularity;
+    }
+
+    result = cuMemCreate(&handle, size, &prop, 0);
+    if (result != CUDA_SUCCESS) {
+        /**
+         * сuMemCreate fails with an error if cannot create the CUDA memory
+         * handle with CU_MEM_HANDLE_TYPE_FABRIC handle type
+         */
+        alloc_mem.ptr = 0;
+        goto out;
+    }
+
+    CUDA_CALL(cuMemAddressReserve(&ptr, size, 0, 0, 0));
+    CUDA_CALL(cuMemMap(ptr, size, 0, handle, 0));
+
+    access_desc.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+    access_desc.location.id   = cu_dev;
+    access_desc.flags         = CU_MEM_ACCESS_FLAGS_PROT_READWRITE;
+    CUDA_CALL(cuMemSetAccess(ptr, size, &access_desc, 1));
+
+    alloc_mem.ptr  = ptr;
+    alloc_mem.size = size;
+    alloc_mem.obj  = (void*)handle;
+
+out:
+    return alloc_mem;
+}
+
+static alloc_mem_t alloc_vmm(size_t size)
+{
+    return alloc_vmm_type(size, 0);
+}
+
+static void free_vmm(alloc_mem_t *alloc_mem)
+{
+    CUDA_CALL(cuMemUnmap(alloc_mem->ptr, alloc_mem->size));
+    CUDA_CALL(cuMemAddressFree(alloc_mem->ptr, alloc_mem->size));
+    CUDA_CALL(cuMemRelease((CUmemGenericAllocationHandle)alloc_mem->obj));
+}
+
+#if HAVE_CUDA_FABRIC
+static alloc_mem_t alloc_vmm_fabric(size_t size)
+{
+    return alloc_vmm_type(size, CU_MEM_HANDLE_TYPE_FABRIC);
+}
+
+static alloc_mem_t alloc_mempool(size_t size)
```

Comment:
can we have mempool without FABRIC?
Or do we have FABRIC cap on our CI machines?

Reply:
I don't think that we need mempool without fabric feature in this test, since it won't check any code path, that isn't covered by existing allocators.

Reply:
is FABRIC currently checked on our CI machines?

Reply:
No, we don't have FABRIC capability on our CI machines, and it is not checked there. Mempool allocator acts like a mempool without fabric handle type.

Reply:
but we define mempool with just FABRIC handles, no? How do we check it without fabric?

Reply:
Yes. The handle type is ignored in our CI.